pynitrokey/cli/update.py

```diff
@@ -45,7 +40,42 @@
 def update(serial, yes, local_firmware_server, alpha):
     """Update Nitrokey key to latest firmware version."""
 
-    update_url = 'https://update.nitrokey.com/'
-    print('Please use {} to run the firmware update'.format(update_url))
-    return
+    #update_url = 'https://update.nitrokey.com/'
+    #print('Please use {} to run the firmware update'.format(update_url))
+    #return
+
+
+    latest_release_url = "https://github.com/Nitrokey/nitrokey-fido2-firmware/releases/latest"
+
+    web_data = requests.get(latest_release_url)
+
+    res = re.findall(r'a href="([^"]+)"', web_data.text)
+    _url = None
+    for item in res:
+        if "firmware" in item and "download" in item and item.endswith(".json"):
+            _url = item
+            break
+    download_url = f"https://github.com{_url}"
+    tmp_dir = tempfile.gettempdir()
+    fw_fn = os.path.join(tmp_dir, "fido2_firmware.json")
+    with open(fw_fn, "wb") as fd:
+        firmware = requests.get(download_url)
+        fd.write(firmware.content)
+
+    print("entering bootloader mode - please confirm by touching the device's button")
+    program.commands["aux"].commands["enter-bootloader"].callback(None)
+
+    program.commands["aux"].commands["bootloader-version"].callback(None)
+
+    print("updating Nitrokey FIDO2 using bootloader mode")
+    program.commands["bootloader"].callback(None, fw_fn)
+
+    # ensure that we are not stuck in bootloader mode...
+    try:
+        program.commands["aux"].commands["leave-bootloader"].callback(None)
+        print("had to leave bootloader explicitly, please check firware version:")
+        print("$ nitropy fido2 verify")
+    except CtapError as e:
+        pass
```

Comment:
While this works, it has major advantages:
- blocks refactoring of the CLI commands;
- makes invisible connections, which might be problematic during code refactoring;
- exceptions/error handling is opaque, might even pass further on error to another commands, making debugging of user scenarios hard.

Please use the code which was here earlier (see `git blame`; originally commented out, and later on deleted on further refactoring / cleanups).

Reply:
yeah was thinking about the same points.
Nevertheless, explicitly choose to go through the other commands, as I wanted to avoid two implementations for semantically equivalent operations. But getting the point, will fix it...